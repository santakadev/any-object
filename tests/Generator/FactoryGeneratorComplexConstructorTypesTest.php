<?php

declare(strict_types=1);

namespace Santakadev\AnyObject\Tests\Generator;

use ApprovalTests\Approvals;
use Santakadev\AnyObject\Tests\Generator\Generated\AnyNamedConstructorWithPrivateConstructObject;
use Santakadev\AnyObject\Tests\Generator\Generated\AnySharedTypesInConstructorObject;
use Santakadev\AnyObject\Tests\Generator\Generated\AnyVariadicNamedConstructorObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\NamedConstructorWithPrivateConstructObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\SharedTypesInConstructorObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\VariadicNamedConstructorObject;

class FactoryGeneratorComplexConstructorTypesTest extends FactoryGeneratorTestCase
{
    public function test_generator_from_named_constructor(): void
    {
        $this->generateFactoryFor(NamedConstructorWithPrivateConstructObject::class);

        $text = $this->readGeneratedAnyFileFor(NamedConstructorWithPrivateConstructObject::class);
        Approvals::verifyString($text);
        $test = AnyNamedConstructorWithPrivateConstructObject::with(value: "test");
        $this->assertEquals("test", $test->value);
    }

    public function test_generator_from_variadic_named_constructor(): void
    {
        $this->generateFactoryFor(VariadicNamedConstructorObject::class);

        $text = $this->readGeneratedAnyFileFor(VariadicNamedConstructorObject::class);
        Approvals::verifyString($text);
        $test = AnyVariadicNamedConstructorObject::with(value: ["a", "b"]);
        $this->assertEquals(["a", "b"], $test->value);
    }

    public function test_generator_shared_types_in_constructor(): void
    {
        $this->generateFactoryFor(SharedTypesInConstructorObject::class);

        $text = $this->readGeneratedAnyFileFor(SharedTypesInConstructorObject::class);
        Approvals::verifyString($text);
        AnySharedTypesInConstructorObject::build();
    }
}
