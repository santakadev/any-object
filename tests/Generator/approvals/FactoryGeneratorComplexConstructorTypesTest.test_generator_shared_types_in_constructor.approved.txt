<?php

namespace Santakadev\AnyObject\Tests\Generator\Generated;

use Faker\Factory;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\SharedTypesInConstructorObject;
use Santakadev\AnyObject\Tests\TestData\ScalarTypes\StringObject;

final class AnySharedTypesInConstructorObject
{
    public static function with(StringObject|None $value1 = new None(), StringObject|None $value2 = new None(), StringObject|None $value3 = new None()) : SharedTypesInConstructorObject
    {
        if ($value1 instanceof None) {
            $faker = Factory::create();
            $value1 = AnyStringObject::build();
        }
        if ($value2 instanceof None) {
            $faker = Factory::create();
            $value2 = AnyStringObject::build();
        }
        if ($value3 instanceof None) {
            $faker = Factory::create();
            $value3 = AnyStringObject::build();
        }
        return new SharedTypesInConstructorObject($value1, $value2, $value3);
    }
    public static function build() : SharedTypesInConstructorObject
    {
        return self::with();
    }
}
